crates/proto/src/udp/udp_stream.rs

```diff
@@ -21,24 +22,24 @@ use tracing::{debug, warn};
 use crate::xfer::{BufDnsStreamHandle, SerialMessage, StreamReceiver};
 use crate::Time;
 
-/// Trait for UdpSocket
+pub(crate) type UdpCreator<S> = Arc<
+    dyn Send
+        + Sync
+        + (Fn(
+            SocketAddr, // local addr
+            SocketAddr, // server addr
+        ) -> Pin<Box<dyn Send + (Future<Output = Result<S, std::io::Error>>)>>),
+>;
```

Comment:
This seems pretty ugly, why do we need this?

Reply:
It's surely ugly, but without it I don't know how to comply to current UDP DNS behavior. For UDP DNS, we connect through a new UDP socket for each request. Use this closure. Everytime the `NextRandomUdpSocket` requires, it invokes the closure. Another approach I came up with just now is moving the `RuntimeProvider` interface to `proto` crate (since udp related codes are there). What about your opinion?

Reply:
Yeah, IMO it would be better to move `RuntimeProvider` into the proto crate.

Reply:
OK, I will work for that.

Reply:
Hi @djc, I tried to fix the probelm with `RuntimeProvider` as struct member. However, to remove this we need to modify `UdpStream` and `UdpClientStream` much. Since `UdpStream` and `UdpClientStream` is widely used among the project, such an introduction will cause lots of break in other places, and probably other ugly codes. I think although this definition looks scary, it is in fact a simple closure, with two `SocketAddr` as arguments and one `Future` as result. The other parts just make compiler happy. It is also private within crate. After the trial my intention is to keep the "ugly" definition. But I also want to hear about your suggestion.